configs/default/IFRC-JBF7.config

```diff
@@ -0,0 +1,212 @@
+# version
+# Betaflight / STM32F7X2 (S7X2) 4.2.0 May 27 2020 / 13:32:04 (6bada9c5e) MSP API: 1.43
+
+board_name JBF7
+manufacturer_id IFRC
+
+# resources
+resource BEEPER 1 C15
+resource MOTOR 1 C08
+resource MOTOR 2 C06
+resource MOTOR 3 C09
+resource MOTOR 4 C07
+resource MOTOR 5 B06
+resource MOTOR 6 B07
+resource MOTOR 7 B01
+resource MOTOR 8 B00
+resource PPM 1 A03
+resource LED_STRIP 1 A01
+resource SERIAL_TX 1 A09
+resource SERIAL_TX 2 A02
+resource SERIAL_TX 3 B10
+resource SERIAL_TX 4 C10
+resource SERIAL_TX 5 C12
+resource SERIAL_RX 1 A10
+resource SERIAL_RX 2 A03
+resource SERIAL_RX 3 B11
+resource SERIAL_RX 4 C11
+resource SERIAL_RX 5 D02
+resource I2C_SCL 2 B10
+resource I2C_SDA 2 B11
+resource LED 1 C04
+resource SPI_SCK 1 A05
+resource SPI_SCK 2 B13
+resource SPI_SCK 3 B03
+resource SPI_MISO 1 A06
+resource SPI_MISO 2 B14
+resource SPI_MISO 3 B04
+resource SPI_MOSI 1 A07
+resource SPI_MOSI 2 B15
+resource SPI_MOSI 3 B05
+resource CAMERA_CONTROL 1 A00
+resource ADC_BATT 1 C01
+resource ADC_RSSI 1 C00
+resource ADC_CURR 1 C02
+resource SDCARD_CS 1 A04
+resource PINIO 1 C13
+resource PINIO 2 C14
+resource PINIO 3 B08
+resource FLASH_CS 1 B09
+resource OSD_CS 1 B12
+resource GYRO_EXTI 1 A08
+resource GYRO_EXTI 2 B02
+resource GYRO_CS 1 A15
+resource GYRO_CS 2 C03
+
+# timer
+timer A00 AF2
+# pin A00: TIM5 CH1 (AF2)
+timer A03 AF3
+# pin A03: TIM9 CH2 (AF3)
+timer C08 AF3
+# pin C08: TIM8 CH3 (AF3)
+timer C06 AF3
+# pin C06: TIM8 CH1 (AF3)
+timer C09 AF3
+# pin C09: TIM8 CH4 (AF3)
+timer C07 AF3
+# pin C07: TIM8 CH2 (AF3)
+timer B06 AF2
+# pin B06: TIM4 CH1 (AF2)
+timer B07 AF2
+# pin B07: TIM4 CH2 (AF2)
+timer B01 AF2
+# pin B01: TIM3 CH4 (AF2)
+timer B00 AF2
+# pin B00: TIM3 CH3 (AF2)
+timer A01 AF1
+# pin A01: TIM2 CH2 (AF1)
+
+# dma
+dma ADC 3 0
+# ADC 3: DMA2 Stream 0 Channel 2
+dma pin A00 0
+# pin A00: DMA1 Stream 2 Channel 6
+dma pin C08 1
+# pin C08: DMA2 Stream 4 Channel 7
+dma pin C06 0
+# pin C06: DMA2 Stream 2 Channel 0
+dma pin C09 0
+# pin C09: DMA2 Stream 7 Channel 7
+dma pin C07 1
+# pin C07: DMA2 Stream 3 Channel 7
+dma pin B06 0
+# pin B06: DMA1 Stream 0 Channel 2
+dma pin B07 0
+# pin B07: DMA1 Stream 3 Channel 2
+dma pin B01 0
+# pin B01: DMA1 Stream 2 Channel 5
+dma pin B00 0
+# pin B00: DMA1 Stream 7 Channel 5
+dma pin A01 0
+# pin A01: DMA1 Stream 6 Channel 3
+
+# feature
+feature -RX_PARALLEL_PWM
+feature RX_SERIAL
+feature OSD
+feature ESC_SENSOR
+
+# beacon
+beacon RX_LOST
+beacon RX_SET
+
+# serial
+serial 0 64 115200 57600 0 115200
+serial 3 1024 115200 57600 0 115200
+serial 4 2048 115200 57600 0 115200
+
+# vtxtable
+vtxtable bands 1
+vtxtable channels 8
+vtxtable band 1 RACEBAND R FACTORY    0    0 5732 5769 5806 5843    0    0
+vtxtable powerlevels 1
+vtxtable powerlabels 25
```

Comment:
Including the vtxtable in the unified config is a bad idea. This presumes that the user will only have a smartaudio VTX as the table is not compatible with Tramp. Further are iFlight and/or JB going to indemnify Betaflight against legal action and guarantee that this table is legal worldwide?

Reply:
Agreed on this point - this will not work for any Tramp user, as tramp does not support 'FACTORY' bands at all - this will give 27% of VTX users a non-working configuration:
![image](https://user-images.githubusercontent.com/4742747/89829802-45800900-dbaf-11ea-9057-efd297ff66b8.png)

Reply:
We very much want to ship with a working vtxtable for the largest number of users. If the user must change the vtxtable because they have some other vTX, then we can walk them through the process.

If we change from "factory" to "non-factory" then Tramp still won't work since power value is not correct--right? But shipping with SA2.0 compatible vtxtable I think will provide working config for the largest number of users by default.

iFlight should speak to indemnification and risk however I believe we have researched and consider these channels to be acceptable to ship. We understand that BF project prefers to ship with un-configured vtxtable because of potential legal risk, and that we are explicitly changing that decision of our own volition.

Reply:
The problem is that this `vtxtable` won't work for SA either. There's a misunderstanding about how the bands work I think. SA can change channels in two different ways. The first is by band/channel (FACTORY mode). In this mode the vtx is sent a band and channel number like 1 and 4. **The actual frequencies in the `vtxtable` are meaningless and are simply labels for the user.** The vtx decides what actual frequency to set based on its internal mapping of band/channel to frequency. Every SA vtx supports this.

The next mode is "CUSTOM". When the band is defined this way then the actual frequency from the `vtxtable` is sent and the band/channel is irrelevant. So in this case the frequency in the table does matter and is the controlling factor. The problem is that this mode is not supported by many SA "compatible" vtx's. And if it is supported then it's only partially supported in that sending a frequency not in the vtx's internal table will generally be ignored (some shut off, others will go to the lowest frequency, or some other unexpected behavior). Real TBS vtx's support the frequency mode correctly and are the only ones that can be counted on for this behavior.

So the problem with the approach in this target is that to most vtx's band "1" is not race band. It's often BOSCAM A or sometimes FATSHARK band. It just depends on how the vtx manufacturer implemented their internal table. So you might think, "well we'll just make it band 5 - that's race band right?". Well no, not necessarily as the 5th band is not guaranteed to be race band. And to do that you'd need to define the vtxtable with 5 bands and it'd look like this:
```
vtxtable bands 5
vtxtable channels 8
vtxtable band 1 BAND1    1 CUSTOM     0    0    0    0    0    0    0    0
vtxtable band 2 BAND2    2 CUSTOM     0    0    0    0    0    0    0    0
vtxtable band 3 BAND3    3 CUSTOM     0    0    0    0    0    0    0    0
vtxtable band 4 BAND4    4 CUSTOM     0    0    0    0    0    0    0    0
vtxtable band 5 RACEBAND R FACTORY    0    0 5732 5769 5806 5843    0    0
vtxtable powerlevels 1
vtxtable powervalues 0
vtxtable powerlabels 25
```
And this would obviously be very confusing (and still not guaranteed to work). And as stated above you can't choose to make the band "CUSTOM" as that's not supported by many non-TBS SA vtx's.

Just a side note that Tramp protocol **only** supports CUSTOM and setting by frequency and doesn't have a band/channel capability. But even then there are a number of different implementations and the vtx won't always set any arbitrary frequency. Some will only set to actual channels as defined in the vtx internally and will stop transmitting if some other value is set. Matek vtx's work this way for example.

Reply:
Do you have a recommendation for how to provide a working vtxtable for the largest number of users? You can probably infer from the vtxtable submitted the channel set that we consider sufficiently legal for worldwide use. I understand that Betaflight project and devs are not willing to assume liability for this decision. But there's nobody more qualified to provide technical advice.

Reply:
What about:

vtxtable band 1 RACEBAND R CUSTOM 0 0 5732 5769 5806 5843 0 0

Reply:
As @etracer65 explains above, many non-TBS SmartAudio VTX do not support `CUSTOM` - so even this would leave a considerable part of users with a configuration that looks ok but does not work with their hardware.
The problem of providing a working VTXTABLE for any given VTX hardware is really a task that is up to the manufacturer of the hardware. Think of this just like you think of the drivers that are needed to make peripherals that you plug into you PC work - they are bits of code that are specific to the hardware you are connecting, and only the manufacturer will have all of the information required to fully support the hardware. If you buy a new Logitech webcam and plug it into your Lenovo PC, and it does not work, you'll expect Logitech to supply you with a driver, and you won't go and write to Lenovo and demand that they give you a driver for your Logitech device, or will you?

Reply:
@joshuabardwell The unified target config is meant to describe and configure the **hardware** of the flight controller to make it functional. It is **not** meant to set user preferences, branding, or other configuration shortcuts. This is why I'm opposed to the rates preferences being included - they have nothing to do with hardware configuration of the flight controller.

Likewise beyond the problems of it being impossible to supply a "universal" vtxtable as discussed above, it's again against the point of the unified target config. It's not meant to configure peripherals that the user may or may not connect. Additionally the vtx channels, frequencies, power levels, etc. were intentionally removed from Betaflight and replaced with vtxtables requiring user action to provide protection from sanction and possible legal actions from governmental agencies as you're aware. Just looking at recent events with massive fines of Hobbyking and earlier GetFPV reinforce this position. So trying to add this back via a unified target config is really not acceptable and in my opinion and an attempt to backdoor our policies.

Reply:
Understood. I have requested iFlight to remove the vtxtable lines and to remove the custom rates lines from the supplied config.

Regarding "configure hardware of FC" vs. "configure peripherals" one point of confusion is that FC targets typically lay out the Ports tab for the expected use of UARTS, such as serial RX enabled on one UART, SmartPort enabled on another, etc. This was the main reason for submitting two separate targets--one for DJI and one for analog.

There hasn't been any objection so far to delivering the target with a custom OSD configuration (one for DJI one for analog). Is that acceptable or does that also fall under user preference and need to be removed?

Reply:
> Regarding "configure hardware of FC" vs. "configure peripherals" one point of confusion is that FC targets typically lay out the Ports tab for the expected use of UARTS, such as serial RX enabled on one UART, SmartPort enabled on another, etc. This was the main reason for submitting two separate targets--one for DJI and one for analog. To save the user the step of configuring their own Ports tab.

I don't have a problem with this. If the configuration matches the labeling on the board then in my opinion it's part of configuring the hardware. Like if the SBUS pad is connected to UART1 then selecting serial rx for that is logical. Or a pad for smartport telemetry and having that UART set with telemetry enabled. Of course the downside to this is if the user want's to do a different setup it often leads to confusion in them thinking they can't use those pads for some other UART-based feature. So to me it's a toss-up whether doing this adds value or not.

> There hasn't been any objection so far to delivering the target with a custom OSD configuration (one for DJI one for analog). Is that acceptable or does that also fall under user preference and need to be removed?

Personally as a user I wouldn't want the OSD elements predefined. But from a developer perspective I'm a bit more on the fence about this and don't have a strong opinion. At a high level I think it's generally a bad idea to "bake" these into the unified target config however. One reason is that it might be perceived to be a benefit to the user in that they don't have to set them up, they are however **forced to turn them off** if they don't want them. And will have to turn them off anytime the board is reset to defaults The other aspect is that future enhancements to the OSD item position definitions (the parameter values) are planned and what's stored in the unified config may not work as expected in the future. So in that case the user might get an error or unexpected result when applying defaults. This is actually the strongest argument against applying non-hardware definition settings in the unified config. We're aware of the dependence of unified targets on specific configuration settings and therefore are careful about how they're used, changed, etc. We won't be expecting user preference or other general settings in the unified config so this will eventually cause a problem.

Reply:
In order to avoid legal risks, the VTX form has been removed.